pkg/trace/writer/trace.go

```diff
@@ -216,8 +216,26 @@ func (w *TraceWriter) flush() {
 		gzipw.Write(b)
 		gzipw.Close()
 
-		for _, sender := range w.senders {
-			sender.Push(p)
+		var payloads []*payload
+		if len(w.senders) == 1 {
+			// Avoid an allocation when only one endpoint is configured.
+			payloadsArr := [1]*payload{p}
+			payloads = payloadsArr[:]
+		} else {
+			// If there is more than one sender then we need to create a payload clone for each
+			// one because pooling is managed at the per-sender level so we need separate payloads
+			// for each sender to avoid double-put pooling bugs.
```

Comment:
The "_If there is more than one sender_" part can be left out because the code makes that obvious. I would also recommend not to mention the reasoning behind the change in this PR (the bug) because people reading the code won't read it from this context. Maybe a smaller explanation will suffice?

```go
// create a clone for each payload because each sender places
// payloads back onto the pool after they are sent.
```